annotations/relations.py

```diff
@@ -152,46 +156,102 @@ def build_dependency_graph(template_data, part_data, **kwargs):
+                raise InvalidTemplate(f"Invalid node format: {node}. Expected format is a number followed by 's', 'p', or 'o'.")
+                
+    except InvalidTemplate:
+        raise
+    except Exception:
         raise InvalidTemplate("Invalid pattern for terminal nodes")
 
 
 def validate_expression(template_data, part_data, **kwargs):
-    N_parts = len(part_data)
-
+    """
+    Validate the format of the expression.
+    Each key in the expression should be two characters: 
+    a digit representing part ID followed by a character 
+    representing field (s, p, or o).
+    
+    No longer validates against terminal nodes.
+    """
+    
     try:
-        keys = list(zip(*list(Formatter().parse(template_data.get('expression')))))[1]
-        for part_id, pred_flag in map(tuple, keys):
-            if not int(part_id) <= N_parts:
-                raise InvalidTemplate("Part ID in expression is invalid.")
-            if not pred_flag in ['s', 'p', 'o']:
-                raise InvalidTemplate("Node ID in expression is invalid.")
-    except ValueError as E:
-        # Raised if there are not precisely two characters in each key.
-        raise InvalidTemplate("Each key in the expression must be precisely"
-                              " two characters long")
+        formatter = Formatter()
+        # Check that keys in the expression have valid syntax
+        for _, key, _, _ in formatter.parse(template_data.get('expression', '')):
+            if key is not None:
+                if len(key) < 2:
+                    raise InvalidTemplate("Each key in the expression must be at least two characters long")
+                
+                try:
+                    # Check first character is a digit
+                    int(key[0]) 
+                    # Check last character is s, p, or o
+                    if key[-1] not in ['s', 'p', 'o']:
+                        raise InvalidTemplate(f"Invalid field identifier in expression key: {key}")
+                except ValueError:
+                    raise InvalidTemplate(f"Invalid format in expression key: {key}")
+                
+    except InvalidTemplate:
+        raise
     except Exception as E:
         raise InvalidTemplate("Invalid expression pattern")
 
 
 def validate_template_data(template_data, part_data, **kwargs):
-    validate_terminal_nodes(template_data, part_data)
-    validate_expression(template_data, part_data)
-
-
+    
+    # Validate terminal nodes
+    if 'terminal_nodes' in template_data and template_data['terminal_nodes']:
+        try:
+            node_list = template_data['terminal_nodes'].split(',')
+            for node in node_list:
+                node = node.strip()
+                if not node:
+                    continue
+                if len(node) < 2 or not node[0].isdigit() or node[-1] not in ['s', 'p', 'o']:
+                    raise InvalidTemplate(f"Invalid node format in terminal_nodes: {node}")
```

Comment:
I know this is the content of the original validation function (which should not copied here but called, see comment below), but this doesn't seem completely right.
1: what if someone enters `,1o`? that would be an incorrect pattern but it would be caught, would it?
2: what if someone enters `1oo`, this would be an incorrect pattern that would not be caught either, would it?